Test2sm

moeller0 opened the pull request:
Hi Hannu,

this is my new attempt at craeting a PR, let's see whether this worked... And you where right, I had started from too old a base, not sure why, but I certainly need a bit more testing, I believe, so the actual patch here is not worth merging, I just wanted a proper test vehicle. ;)

Thanks for you help...

hnyman commented:
Again there are ~30 extra comments in the PR (  0080543- f875bf3). There should only be your commit (28d806c).

I think that the reason is that you have pulled all very recent commits from openwrt/packages to the test2sm branch from which you are submitting the PR against hnyman/packages. Several pull requests were accepted yesterday in openwrt/packages and I have not refreshed/updated my github fork with those changes, yet, (while you have pulled them into your local copy and are now pushing them to my repo).

In general, when creating a pull request, you should have in that branch just the commits that are in the repo to which you are committing (plus your own changes). You should not pull extra commits from some third place. This is likely no problem when you are working against "the upmost parent" repo openwrt/packages, but my hnyman/packages is a fork that I have to manually update (pull upstream / push) after changes in openwrt/packages. So, when working with it, updating your copy further from the parent than my repo, can lead to confusion.

I will try pulling the new commits from openwrt/packages to hnyman/packages and will see if that corrects how your PR is shown to me.

hnyman commented:
I refreshed my fork master by "git pull upstream master" to get new commits from openwrt/packages and "git push" to update them in github copy hnyman/packages. (But that pulled also 8 new commits made after you created your PR, so now master in hnyman/packages is 8 commits further ahead than the starting point of your PR.)

But I still see the 30 extra commits in your PR's listing of new commits. It needs to be refreshed somehow.

After my update, the history is already shown rationally in Guthub GUI. So Github correctly sees that there is actually only one new commit from you.
See https://github.com/hnyman/packages/network
(the hnyman/packages master is in the top, and scroll down to see your commit. your own commit starts from the correct parent and only one commit is from you. So in principle Github sees it correctly.)

I think that a "git rebase" will fix this. It would certainly do it if the test2sm branch would only contain your own commit, but there are the 30 extra commits shown in the PR, so I am not quite sure.

It might also be that I could just accept the change and github would overlook the duplicate commits shown in the PR listing. But I would not like to try that, as the PR should not show/include any extra commits.

Below are two different fix attempts. Please try them. (If the first one works, the second one is not needed.) I am interested to hear the result, as this kind of error may surface also with others at some point.

1)
It might be enough if you just refresh your local git log and then force push the PR branch again (as github sees the branching point correctly, but the PR shows it wrongly as it is based on old information.) That simple refresh might update the branching point to the correct location. Try updating master (to get new git status info and new master commits):

```
git checkout master
git pull
git checkout test2sm
git push -f
```

(That should update the PR. You can add and/or overwrite the PR branch. the PR get updated by that.)

Actually, already before "push -f" you should see the correct branching with `git log --oneline --graph`.

2)
If the first simple fix fails, try this one:
"git rebase" is the command to use when you have a branch with your own commits and you want to update it to the tip/HEAD of the master that contains new commits created after your branch. Work flow goes like this: update your local master branch, checkout your new branch, try to rebase it to start from the current HEAD instead of old HEAD at branch creation moment. Git will try to insert new commits from master one by one into your branch before your personal commits. If it succeeds, your branch has been transferred to start from the current HEAD.

```
git checkout master
git pull  (to fully update your local master branch)
git checkout test2sm
git rebase master    (this is the actual fix, rebase current branch against master)
git push -f    (to push into Github and to update PR)
```

Similarly also here, `git log --oneline --graph` should show the correct commits after git rebase.

moeller0 commented:
Hi Hannu,

again thanks for all your help. I promise I will delre that repository and start fresh for openwrt/packages. But foe testing, I believe my current repository seems quite useful.

> On Jun 14, 2016, at 09:09 , Hannu Nyman notifications@github.com wrote:
> 
> I refreshed my fork master by "git pull upstream master" to get new commits from openwrt/packages and "git push" to update them in github copy hnyman/packages. (But that pulled also 8 new commits made after you created your PR, so now master in hnyman/packages is 8 commits further ahead than the starting point of your PR.)
> 
> But I still see the 30 extra commits in your PR's listing of new commits. It needs to be refreshed somehow.
> 
> After my update, the history is already shown rationally in Guthub GUI. So Github correctly sees that there is actually only one new commit from you.
> See https://github.com/hnyman/packages/network
> (the hnyman/packages master is in the top, and scroll down to see your commit. your own commit starts from the correct parent and only one commit is from you. So in principle Github sees it correctly.)
> 
> I think that a "git rebase" will fix this. It would certainly do it if the test2sm branch would only contain your own commit, but there are the 30 extra commits shown in the PR, so I am not quite sure.
> 
> It might also be that I could just accept the change and github would overlook the duplicate commits shown in the PR listing. But I would not like to try that, as the PR should not show/include any extra commits.
> 
> Below are two different fix attempts. Please try them. (If the first one works, the second one is not needed.) I am interested to hear the result, as this kind of error may surface also with others at some point.
> 
> 1)
> It might be enough if you just refresh your local git log and then force push the PR branch again (as github sees the branching point correctly, but the PR shows it wrongly as it is based on old information.) That simple refresh might update the branching point to the correct location. Try updating master (to get new git status info and new master commits):
> 
> git checkout master
> git pull
> git checkout test2sm
> git push -f

Okay:
bash-3.2$ git branch
  master
  sqm-extra
  test
- test2sm
  wip
  bash-3.2$ git checkout master
  Switched to branch 'master'
  Your branch is up-to-date with 'origin/master'.
  bash-3.2$ git pull
  Already up-to-date.
  bash-3.2$ git checkout test2sm
  Switched to branch 'test2sm'
  Your branch is up-to-date with 'origin/test2sm'.
  bash-3.2$ git push -f
  Everything up-to-date

That seems to have not changed much/anything? moeller/packages claims to be 9 commits behind hnyman:master and the PR in your repository still contains more than just my commit…
Now I had some issues, do to issuing “git checkout master ; git branch -D test2sm ; git fetch origin ; git merge origin/test2sm” (I deleted the branch by accident and then when I wanted to recreate it from origin I accidentally merged it into master, after resetting master and “git checkout -b test2sm ; git fetch origin ; git merge origin/test2sm” I believe I am back to the status before my accident.)

> (That should update the PR. You can add and/or overwrite the PR branch. the PR get updated by that.)
> 
> Actually, already before “push -f" you should see the correct branching with git log --oneline --graph.

```
Okay that still looks larger than my single commit…
```

> 2)
> If the first simple fix fails, try this one:
> "git rebase" is the command to use when you have a branch with your own commits and you want to update it to the tip/HEAD of the master that contains new commits created after your branch. Work flow goes like this: update your local master branch, checkout your new branch, try to rebase it to start from the current HEAD instead of old HEAD at branch creation moment. Git will try to insert new commits from master one by one into your branch before your personal commits. If it succeeds, your branch has been transferred to start from the current HEAD.
> 
> git checkout master
> git pull  (to fully update your local master branch)
> git checkout test2sm
> git rebase master    (this is the actual fix, rebase current branch against master)
> git push -f    (to push into Github and to update PR)

bash-3.2$ git checkout master
Switched to branch 'master'
Your branch is up-to-date with 'origin/master'.
bash-3.2$ git pull
Already up-to-date.
bash-3.2$ git checkout test2sm
Switched to branch 'test2sm'
bash-3.2$ git rebase master 
Current branch test2sm is up to date.
bash-3.2$ git push -f 
Everything up-to-date
bash-3.2$ 

> Similarly also here, git log —oneline --graph should show the correct commits after git rebase.

```
This still is full of extra commits…
```

But I believe I found the issue I had “git remote add upstream git@github.com:openwrt/packages.git” osed openwrt/packages as upstream instead of hnyman/packages. But even adding your repository as remote still gives a weird “git log —oneline —graph” even though the github PR it looks saner now, does that mean that the rebase has worked?
    I guess it is time to delete that repository and start fresh…

Best Regards
    Sebastian

> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub, or mute the thread.

hnyman commented:
Now it looks ok to me. PR has just one commit and it starts from a sensible place in the network graph. 
https://github.com/hnyman/packages/network
(please note that you may need to switch to the "network" graph tab 5-6 times before Github updates it. Github is lazy about that.)

> But I believe I found the issue I had “git remote add upstream git@github.com:openwrt/packages.git” osed openwrt/packages as upstream instead of hnyman/packages. 

Yeah, as you cloned from my repo, you have local repo + three upstream repos: "origin" is your own fork that you can write to and you need to keep updated, hnyman/packages (the actual parent, against which you submit pull requests, you can name that "upstream" or "hnyman" or whatever, as long as you keep in my that this is the actual upstream for you) and in the background there is also openwrt/packages that you should not actually pull from if you want to submit anything to hnyman/packages.

For the actual maintenance work you should clone a new repo directly from openwrt/packages, then it is much clearer.

I think that I actually gave you wrong instructions earlier.
Plain "git pull" pulls from origin, i.e. your own github fork master, but you need actually "git pull upstream master" (assuming that upstream = hnyman/packages). Sorry about that.

The normal "refresh master from upstream" action goes like:
git checkout master
git pull upstream master  (new commits and branches from upstream to local repo)
git push (to push the possible new commits from local repo to your github fork "origin")

After that your local master and github fork master are up-to-date with the upstream master.

hnyman commented:
You could also testing ading a second commit to the PR.
Just commit something to the same branch and then "git push"  (no -f needed).
Then you should see both commits in the PR.

Then you could tidy things up and "squash" the two commits into one with interactive rebase.
git checkout test2sm
git rebase -i HEAD~2   (to open the rebase editor with 2 last commits)
(then edit the second line to have "s" or "squash" as the first word, and close editor. That will combine the second commit into the first one and open and editor to edit the combined commit message)
git push -f  (to update the github form and PR along it. After this there shold again be just one commit in the PR branch and Github GUI)

That is a very neat way to tidy private history before sending permanently into the public main repo. Especially useful if development is messy or there are changes requested for a PR (add/edit commit message, change a line etc.). I just wish that cake & tc-adv repos would have been handled this way. Jonathan & Dave should have squashed their daily work in the local repo and committed one clean commit for each feature change. (It makes no sense to have 4-5 minor correction commits to the earlier commits of the same day...)

Interactive rebase example with 3 commits. I tell git to use first commit as it is, but to combine both 2 and 3 to the first one. Commits could also be dropped, just the commit message changed etc. After I close this rebase editor, git will open a new editor enabling me to edit the combined commit message of these three commits (or actually text from 1 (pick, the main commit) and 2 (squash, use commit and message), as I tell git to fixup 3 (use it, drop commit message):

```
git rebase -i HEAD~3
...in editor:
pick af3fd43 Create users_overview.htm
s af6c778 Updated user.png
f 939fa2f fix uci-defaults

# Rebase 12c3e37..939fa2f onto 12c3e37 (3 command(s))
#
# Commands:
# p, pick = use commit
# r, reword = use commit, but edit the commit message
# e, edit = use commit, but stop for amending
# s, squash = use commit, but meld into previous commit
# f, fixup = like "squash", but discard this commit's log message
# x, exec = run command (the rest of the line) using shell
# d, drop = remove commit
#
```

This advice might have been unnecessary for you, but I am not quite sure how much you know about git. I am writing this as my own first public pull request got rather messy when changes and squash were requested and I started merging etc. I had then not yet understood how neatly Github PRs allow us to edit/overwrite the PR commits ;-)

hnyman commented on net/sqm-scripts-extra/src/test_triple_isolated_llt_cake.qos.help:
```diff
@@ -1,3 +1,3 @@
 Cake qdisc with triple-isolate and diffserv-llt (Latency-Loss-Tradeoff)
-options as the shaper and fq_codel as leaf qdisc. This script requires
+options as the shaper and cake as leaf qdisc. This script requires
```
Just testing the line-base comment feature. Pretty easy to point out the place where something needs to be changed in the PR.
